### ISSUE TITLE
`TypedDict` error for non-required key despite guarded access

### ISSUE THREAD

**Issue by felixvd** [microsoft/pyright#5096]
**Describe the bug**
When accessing a `TypedDict` field after checking `if d.get('field'):`, a `reportTypedDictNotRequiredAccess` error is shown although the key has to exist after the check.

**To Reproduce**
Run Pyright on this code:

```python
from typing import TypedDict
myDictType = TypedDict("myDictType", {'field': str}, total=False)

def fn(d: myDictType):
    if d.get('field'):
        print(d['field'])  # Error: reportTypedDictNotRequiredAccess
    if 'field' in d:
        print(d['field'])

    # Workaround
    if 'field' in d and d['field'] is not None:
        print(d['field'])
```

Explicitly checking `if d.get('field') is not None:` has the same effect.

**Expected behavior**
Both `if` conditions should be considered valid access guards and suppress the check for the key.

**VS Code extension or command-line**
Pylance v2023.1.10 (VS Code)

**Additional context**
Our code base has several instances where the return of `.get()` is used as access guard and value comparison in one. It would be nice if this error did not appear. The workaround in the code snippet is longer and also slower if the field exists in the dictionary. 

Of the [`TypedDict`-related issues](https://github.com/microsoft/pyright/issues?q=is%3Aissue+access+TypedDict+is%3Aclosed), https://github.com/microsoft/pyright/issues/4887 and https://github.com/microsoft/pyright/issues/5056 seem to be the closest ones. https://github.com/microsoft/pyright/issues/1739 turned up when I tried to Google this.

**Comment by erictraut** [comment]
This isn't a type guard pattern that pyright currently supports. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

Each type guard pattern requires specific code to detect and handle that specific pattern and adds to type evaluation overhead (since it's in the performance-critical code flow evaluation path), so we are careful about adding support for new patterns. We do occasionally add support for new patterns, but I would be reluctant to add support for this one because it involves hard-coding knowledge of a particular method (`get`) in a particular class (`dict`), and that's something we generally try to avoid in type checker logic.

I recommend changing your code to use the `'field' in d` type guard pattern, which is already supported by pyright. The other workaround that you identified would also work.
